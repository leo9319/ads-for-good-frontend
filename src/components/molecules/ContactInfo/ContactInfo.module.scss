@import '../../../styles/colours';
@import '../UserProfile/UserProfile.scss';

.form {
  display: inline-grid;
  gap: var(--spacing-lg, 24px);
  width: 100%;

  .btnGroup {
    @extend .cancelAndSaveBtn;
  }

  .formLayout,
  .formSection {
    display: flex;
    flex-direction: column;
    align-items: flex-start;
    align-self: stretch;
  }

  .formLayout,
  .formSection {
    gap: var(--spacing-sm, 12px);
  }

  .formSection {
    .input {
      width: 100%;
    }
  }

  .headingSection {
    font-weight: var(--typography-font-weight-semibold) !important;
    font-size: var(--typography-font-size-4) !important;
    line-height: var(--typography-line-height-4) !important;
  }

  .subHeadingSection {
    font-weight: var(--typography-font-weight-regular) !important;
    font-size: var(--typography-font-size-2) !important;
    line-height: var(--typography-line-height-2) !important;
    color: $color-neutral-600;
  }

  .editBtn {
    @extend .userProfileEditBtn;
  }
}
