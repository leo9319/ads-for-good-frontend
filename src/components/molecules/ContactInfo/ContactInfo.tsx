import React, { ReactElement, useEffect, useState } from 'react';
import { useForm, FieldValues } from 'react-hook-form';
import Button from '@radix-styles/atoms/Button';
import Input from '@components/atoms/Input';
import { FormErrorProps, payloadProps } from '@internal/types/common';
import ButtonGroup from '@radix-styles/molecules/ButtonGroup';
import { ButtonProps } from '@radix-styles/atoms/Button';
import Heading from '@radix-styles/atoms/Heading';
import Text from '@radix-styles/atoms/Text';
import Box from '@radix-styles/atoms/Box';
import SmallCard from '../Cards/SmallCard';
import styles from './ContactInfo.module.scss';
import { AnalyticsHandler } from '@utils/Analytics';

/**
 * @interface
 * The Contact Form props type of {@link ContactInfo}.
 */

export interface ContactInfoProps {
  /**
   * Title of the Form
   * @property
   */
  title: string;
  /**
   * An optional field that is used to show the info to the user
   * @property
   */
  description: string;
  /**
   * Placeholder for Email input for the login screen
   * @property
   */
  primaryPhonePlaceholder?: string;
  /**
   * Placeholder for Password input for the login screen
   * @property
   */
  alternatePhonePlaceholder?: string;
  /**
   * Pass button text label, if need to display.
   * @property
   */
  text: string;
  /**
   * Pass button text label, if need to display.
   * @property
   */
  textSave: string;
  /**
   * Pass button text label, if need to display.
   * @property
   */
  textCancel: string;
  /**
   * Function to handle form submission
   * @property
   */
  onSubmit: (data: FieldValues) => void;
  /**
   * Errors that will be shown if any occur
   * @property
   */
  formError?: FormErrorProps;
  /**
   * It is to enable and disable the Spinner.
   * @property
   */
  loading?: boolean;
  /**
   * Handling the form data in an object.
   * @property
   */
  formData?: { primaryPhone: string; alternatePhone: string };
  /**
   * Status of the form submission.
   * @property
   */
  formSubmitStatus?: boolean;
}

export const ContactInfo = ({
  title,
  description,
  text,
  textSave,
  textCancel,
  onSubmit,
  formError,
  loading = false,
  primaryPhonePlaceholder = '',
  alternatePhonePlaceholder = '',
  formData = { primaryPhone: '', alternatePhone: '' },
  formSubmitStatus = false,
}: ContactInfoProps): ReactElement => {
  const {
    register,
    handleSubmit,
    reset,
    getValues,
    setValue,
    setError,
    formState: { errors },
  } = useForm<FieldValues>({
    mode: 'onBlur',
  });

  const [status, setStatus] = useState<number | undefined>(undefined);
  const [editable, setEditable] = useState<boolean | undefined>(false);

  const [prevPrimaryPhone, setPrevPrimaryPhone] = useState<string | undefined>(
    formData.primaryPhone
  );
  const [prevAlternatePhone, setPrevAlternatePhone] = useState<
    string | undefined
  >(formData.alternatePhone);

  const { trackClick } = AnalyticsHandler();

  useEffect(() => {
    if (formError?.field && formError.message) {
      reset(getValues());
      setError(formError.field, {
        type: 'server',
        message: formError.message,
      });
      setStatus(formError.status);
    }
  }, [formError, reset, setError]);

  useEffect(() => {
    setPrevPrimaryPhone(formData.primaryPhone);
    setPrevAlternatePhone(formData.alternatePhone);
    convertToFormatedPhoneNumber('primaryPhone', formData.primaryPhone);
    convertToFormatedPhoneNumber('alternatePhone', formData.alternatePhone);
  }, [formData]);

  useEffect(() => {
    if (formSubmitStatus) {
      setEditable(false);
    }
  }, [formSubmitStatus]);

  const convertToFormatedPhoneNumber = (fieldName: string, number: string) => {
    let formatted = '';
    if (number) {
      const value = number.replace(/\D/g, '');
      if (value.length > 0) {
        formatted += '(' + value.substring(0, 3);
      }
      if (value.length > 3) {
        formatted += ') ' + value.substring(3, 6);
      }
      if (value.length > 6) {
        formatted += '-' + value.substring(6, 10);
      }
    }
    return setValue(fieldName as string, formatted.trim() as string);
  };

  let buttonObj: Array<ButtonProps> = [
    {
      text: textCancel,
      type: 'reset',
      size: 'md',
      mode: 'secondary',
      disabled: loading,
      className: styles.btnGroup,
      onClick: () => {
        setEditable(false);
        convertToFormatedPhoneNumber(
          'primaryPhone',
          prevPrimaryPhone as string
        );
        convertToFormatedPhoneNumber(
          'alternatePhone',
          prevAlternatePhone as string
        );
        reset(getValues());
        trackClick({
          target: '',
          location: 'Contact',
          name: 'Contact-Cancel',
          type: 'button',
          contentTypeOrPosition: '',
        });
      },
    },
    {
      text: textSave,
      type: 'submit',
      size: 'md',
      mode: 'tertiary',
      disabled: loading,
      loading: loading,
      className: styles.btnGroup,
    },
  ];

  const formSubmit = (data: payloadProps) => {
    setPrevPrimaryPhone(String(data.primaryPhone));
    setPrevAlternatePhone(String(data.alternatePhone));
    if (onSubmit) {
      onSubmit(data);
    }
  };

  const handleEdit = () => {
    setEditable(true);
    trackClick({
      target: '',
      location: 'Contact',
      name: 'Contact-Edit',
      type: 'button',
      contentTypeOrPosition: '',
    });
  };

  return (
    <SmallCard size="3" mode="inverse">
      <form onSubmit={handleSubmit(formSubmit)} className={styles.form}>
        <Heading className={styles.headingSection}>{title}</Heading>
        <Box className={styles.formLayout}>
          <Box className={styles.formSection}>
            <Text className={styles.subHeadingSection} align="left">
              {description}
            </Text>
            <Input
              label="Primary Phone"
              name="primaryPhone"
              className={styles.input}
              disabled={!editable}
              placeholder={primaryPhonePlaceholder}
              rules={{
                required: 'Primary Phone No. is required',
                minLength: {
                  value: 14,
                  message:
                    'Enter a valid phone number with the area code, no special characters.',
                },
              }}
              register={register}
              error={errors}
              status={status}
              type="text"
              onChange={e =>
                convertToFormatedPhoneNumber('primaryPhone', e.target.value)
              }
              readOnly={!editable}
            />
            <Input
              label="Alternate Phone (Optional)"
              name="alternatePhone"
              className={styles.input}
              placeholder={alternatePhonePlaceholder}
              rules={{
                minLength: {
                  value: 14,
                  message:
                    'Enter a valid phone number with the area code, no special characters.',
                },
              }}
              register={register}
              error={errors}
              status={status}
              disabled={!editable}
              type="text"
              onChange={e =>
                convertToFormatedPhoneNumber('alternatePhone', e.target.value)
              }
              readOnly={!editable}
            />
          </Box>
        </Box>
        {editable ? (
          <ButtonGroup className={styles.nextBtnContainer} data={buttonObj} />
        ) : (
          <Button
            type="button"
            text={text}
            className={styles.editBtn}
            mode="secondary"
            onClick={handleEdit}
            iconLeft={true}
            icon="icon-edit"
          />
        )}
      </form>
    </SmallCard>
  );
};

export default ContactInfo;
