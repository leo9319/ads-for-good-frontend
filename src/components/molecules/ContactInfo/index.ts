export type { ContactInfoProps } from './ContactInfo';
export { default } from './ContactInfo';
export { ContactInfo } from './ContactInfo';
