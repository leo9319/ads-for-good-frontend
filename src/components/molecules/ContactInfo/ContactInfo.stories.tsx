import React from 'react';
import ContactInfo from './ContactInfo';
import { Meta, StoryObj } from '@storybook/react';
import Theme from '@radix-styles/atoms/Theme';

const meta = {
  title: 'Library/Molecules/ContactInfo',
  component: ContactInfo,
  decorators: [
    Story => (
      <Theme>
        <Story />
      </Theme>
    ),
  ],
  parameters: {
    layout: 'fullscreen',
  },
  tags: ['autodocs'],
  argTypes: {
    primaryPhonePlaceholder: {
      control: { type: 'text' },
    },
    alternatePhonePlaceholder: {
      control: { type: 'text' },
    },
    formData: {
      control: { type: 'object' },
    },
    text: {
      control: { type: 'text' },
    },
    textSave: {
      control: { type: 'text' },
    },
    textCancel: {
      control: { type: 'text' },
    },
    title: {
      control: { type: 'text' },
    },
    description: {
      control: { type: 'text' },
    },
  },
  args: {
    title: 'Contact',
    description: 'Phone numbers',
    text: 'Edit',
    textSave: 'Save',
    textCancel: 'Cancel',
  },
} satisfies Meta<typeof ContactInfo>;

export default meta;

type Story = StoryObj<typeof meta>;

export const ContactForm: Story = {
  args: {
    primaryPhonePlaceholder: 'Primary Phone No.',
    alternatePhonePlaceholder: 'Alternate Phone No.',
    formData: { primaryPhone: '1234567890', alternatePhone: '' },
    onSubmit: data => console.log(data),
  },
};
