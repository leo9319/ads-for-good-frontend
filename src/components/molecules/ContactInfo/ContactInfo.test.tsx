import React from 'react';
import { render } from '@testing-library/react';
import ContactInfo from './ContactInfo';

describe('ContactInfo', () => {
  it('Should render successfully', () => {
    const { baseElement } = render(
      <ContactInfo
        title="Contact"
        description="Phone numbers"
        text="Edit"
        textSave="Save"
        textCancel="Cancel"
        formData={{ primaryPhone: '1234567890', alternatePhone: '' }}
        onSubmit={data => console.log(data)}
      />
    );
    expect(baseElement).toMatchSnapshot();
  });
});
