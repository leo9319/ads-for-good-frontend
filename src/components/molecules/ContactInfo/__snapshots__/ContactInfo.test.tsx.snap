// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContactInfo Should render successfully 1`] = `
<body>
  <div>
    <div
      class="rt-reset rt-BaseCard rt-Card rt-r-size-1 rt-variant-surface container size-3 inverse"
    >
      <form
        class="form"
      >
        <div
          class="rt-Box formLayout"
        >
          <h1
            class="rt-Heading rt-r-size-6 headingContainer mocked-class-name typography-font-family-default headingSection"
            id=""
          >
            Contact
          </h1>
          <div
            class="rt-Box formSection"
          >
            <span
              class="rt-Text rt-r-ta-left mocked-class-name typography-font-family-default subHeadingSection"
              id=""
            >
              Phone numbers
            </span>
            <div
              class="formInput"
            >
              <label
                for="primaryPhone"
              >
                Primary Phone
              </label>
              <input
                id="primaryPhone"
                name="primaryPhone"
                placeholder=""
                readonly=""
                type="text"
              />
            </div>
            <div
              class="formInput"
            >
              <label
                for="alternatePhone"
              >
                Alternate Phone
              </label>
              <input
                id="alternatePhone"
                name="alternatePhone"
                placeholder=""
                readonly=""
                type="text"
              />
            </div>
          </div>
        </div>
        <button
          aria-label="Edit"
          class="rt-reset rt-BaseButton rt-r-size-2 rt-variant-solid rt-Button wvButton xl secondary none contactBtn"
          data-accent-color=""
          type="button"
        >
          <div
            class="buttonContent iconLeft"
          >
            <span>
              Edit
            </span>
            <span
              class="icon"
              style="border-radius: 0; display: grid; align-items: center; justify-content: center;"
            >
              <span
                class="icon icon-edit"
              />
            </span>
          </div>
        </button>
      </form>
    </div>
  </div>
</body>
`;
